File: java/CacheLoader/src/main/java/com/alachisoft/ncache/samples/cacheLoaderImpl/Refresher.java
package com.alachisoft.ncache.samples.cacheLoaderImpl;

import com.alachisoft.ncache.client.Cache;
import com.alachisoft.ncache.client.CacheItem;
import com.alachisoft.ncache.client.CacheManager;
import com.alachisoft.ncache.runtime.cacheloader.CacheLoader;
import com.alachisoft.ncache.runtime.cacheloader.RefreshPreference;
import com.alachisoft.ncache.samples.data.Product;
import com.alachisoft.ncache.samples.data.Suppliers;

import java.io.InvalidObjectException;
import java.sql.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Refresher implements CacheLoader {
    private static Cache _cache;
    private static String _connectionString;
    private static String _user;
    private static String _password;
    @Override
    public void init(Map<String, String> parameters, String cacheName) throws Exception {
        if(parameters != null && !parameters.isEmpty())
        {
            _connectionString = parameters.getOrDefault("connString", "");
            _user = parameters.getOrDefault("user", "");
            _password = parameters.getOrDefault("password", "");
        }
        if(cacheName == null || cacheName.isEmpty())
            throw new IllegalArgumentException("Cache name cannot be null or empty.");

        _cache = CacheManager.getCache(cacheName);
    }

    @Override
    public Object loadDatasetOnStartup(String dataset) throws Exception {
        List<Object> loadDatasetAtStartup;
        if(dataset==null||dataset.isEmpty())
            throw new InvalidObjectException("Invalid dataset");

        switch (dataset.toLowerCase())
        {
            case "products":
                loadDatasetAtStartup = FetchProductsFromDatabase();
                break;
            case "suppliers":
                loadDatasetAtStartup = FetchSuppliersFromDatabase();
                break;
            default:
                throw new IllegalArgumentException("Invalid dataset");
        }
        String[] keys = GetKeys(loadDatasetAtStartup);
        Map<String, CacheItem> cacheData = GetCacheItemMap(keys, loadDatasetAtStartup);
        _cache.insertBulk(cacheData);
        return LocalDate.now();
    }

    @Override
    public Object refreshDataset(String dataset, Object userContext) throws Exception {
        if(dataset==null||dataset.isEmpty())throw new IllegalArgumentException("Invalid Dataset.");

        LocalDate lastRefreshTime;
        switch (dataset.toLowerCase())
        {
            case "products":
                lastRefreshTime = (LocalDate) userContext;
                List<Object> productsNeedToRefresh = FetchUpdatedProducts(lastRefreshTime);
                for (Object prod :
                        productsNeedToRefresh) {
                    Product product = (Product) prod;
                    String key = "ProductID:"+product.getId();
                    CacheItem cacheItem = new CacheItem(product);
                    _cache.insert(key, cacheItem);
                }
                break;
            case "suppliers":
                lastRefreshTime = (LocalDate) userContext;
                List<Object> suppliersNeedToRefresh = FetchUpdatedSuppliers(lastRefreshTime);
                for (Object sup :
                        suppliersNeedToRefresh) {
                    Suppliers supplier = (Suppliers)sup;
                    String key = "SupplierID:" + supplier.getId();
                    CacheItem cacheItem = new CacheItem(supplier);
                    _cache.insert(key, cacheItem);
                }
                break;
            default:
                throw new IllegalArgumentException("Invalid Dataset.");
        }
        return LocalDate.now();
    }

    @Override
    public Map<String, RefreshPreference> getDatasetsToRefresh(Map<String, Object> map) throws SQLException {

        Map<String, RefreshPreference> datasetNeedToRefresh = new HashMap<>();
        LocalDate lastRefreshTime;
        boolean datasetHasUpdated;
        for (String key :
                map.keySet()) {
            switch (key.toLowerCase())
            {
                case "products":
                    lastRefreshTime = (LocalDate) map.get(key);
                    datasetHasUpdated = hasProductDatasetUpdated(key, lastRefreshTime);
                    if(datasetHasUpdated){
                        datasetNeedToRefresh.put(key, RefreshPreference.RefreshNow);
                    }
                    break;
                case "suppliers":
                    lastRefreshTime = (LocalDate) map.get(key);
                    datasetHasUpdated = hasSupplierDatasetUpdated(key, lastRefreshTime);
                    if(datasetHasUpdated){
                        datasetNeedToRefresh.put(key, RefreshPreference.RefreshOnNextTimeOfDay);
                    }
                    break;
                default:
                    throw new IllegalArgumentException("Invalid Dataset.");
            }
        }
        return datasetNeedToRefresh;
    }

    private Map<String, CacheItem> GetCacheItemMap(String[] keys, List<Object> values) {
        Map<String, CacheItem> items = new HashMap<>();
        CacheItem cacheItem;

        for(int i=0; i<values.size(); i++)
        {
            cacheItem = new CacheItem(values.get(i));
            items.put(keys[i], cacheItem);
        }
        return items;
    }

    private String[] GetKeys(List<Object> objects) {
        String[] keys = new String[objects.size()];
        for(int i=0; i<keys.length;i++)
        {
            if(objects.get(i) instanceof Product)
            {
                Product product = (Product)objects.get(i);
                keys[i] = "ProductID:" + product.getId();
            }
            else if (objects.get(i) instanceof Suppliers)
            {
                Suppliers suppliers = (Suppliers)objects.get(i);
                keys[i] = "SupplierID:" + suppliers.getId();
            }else keys[i] = "";
        }
        return keys;
    }

    private List<Object> FetchSuppliersFromDatabase() throws SQLException {
        String query = "SELECT * FROM SUPPLIERS";
        return executeQuery(query, "suppliers");
    }

    private List<Object> FetchProductsFromDatabase() throws SQLException {
        String query = "SELECT * FROM PRODUCTS WHERE UNITSINSTOCK > 0";
        return executeQuery(query, "products");
    }

    private List<Object> FetchUpdatedSuppliers(LocalDate lastRefreshTime) throws SQLException {
        String query = "SELECT * FROM PRODUCTS WHERE CREATIONTIME > "
                + lastRefreshTime
                + " LASTMODIFY > "
                + lastRefreshTime;
        return executeQuery(query, "suppliers");
    }

    private List<Object> FetchUpdatedProducts(LocalDate lastRefreshTime) throws SQLException {
        String query = "SELECT * FROM PRODUCTS WHERE CREATIONTIME > "
                + lastRefreshTime
                + " LASTMODIFY > "
                + lastRefreshTime;
        return executeQuery(query, "products");
    }

    private boolean hasSupplierDatasetUpdated(String key, LocalDate lastRefreshTime) throws SQLException {
        String query = "SELECT COUNT(*) FROM SUPPLIERS WHERE CREATIONTIME > "
                + lastRefreshTime
                + " LASTMODIFY > "
                + lastRefreshTime;
        return executeAggregateQuery(query) > 0;
    }

    private boolean hasProductDatasetUpdated(String key, LocalDate lastRefreshTime) throws SQLException {
        String query = "SELECT COUNT(*) FROM PRODUCTS WHERE CREATIONTIME > "
                + lastRefreshTime
                + " LASTMODIFY > "
                + lastRefreshTime;
        return executeAggregateQuery(query) > 0;
    }

    private int executeAggregateQuery(String query) throws SQLException {
        int result = 0;
        Connection _conn;
        Statement _st;
        _conn = DriverManager.getConnection(_connectionString, _user, _password);
        _st = _conn.createStatement();
        ResultSet rs = _st.executeQuery(query);
        while (rs.next())
        {
            result = rs.getInt(1);
        }
        _st.close();
        _conn.close();
        return result;
    }

    private List<Object> executeQuery(String query, String dataset) throws SQLException {
        Connection _conn;
        Statement _st;
        List<Object> data;
        _conn = DriverManager.getConnection(_connectionString, _user, _password);
        _st = _conn.createStatement();

        ResultSet rs = _st.executeQuery(query);
        data = getData(rs, dataset);

        _st.close();
        _conn.close();

        return data;
    }

    private List<Object> getData(ResultSet set, String dataSet) throws SQLException {
        List<Object> dataList = new ArrayList<>();
        while (set.next())
        {
            if(dataSet.equals("suppliers"))
            {
                Suppliers suppliers = new Suppliers();
                suppliers.setId(set.getInt(1));
                suppliers.setContactName(set.getString(2));
                suppliers.setCompanyName(set.getString(3));
                suppliers.setAddress(set.getString(4));
                dataList.add(suppliers);
            }else if(dataSet.equals("products"))
            {
                Product product = new Product();
                product.setId(set.getInt(1));
                product.setName(set.getString(2));
                product.setUnitsAvailable(set.getInt(3));
                product.setUnitPrice(set.getInt(4));
                dataList.add(product);
            }
        }
        return dataList;
    }

    @Override
    public void close() throws Exception {

    }
}

File: dotnetcore/RefresherSample/RefresherSample/Refresher.cs
using Alachisoft.NCache.Client;
using Alachisoft.NCache.Runtime.CacheLoader;
using Alachisoft.NCache.Sample.Data;

using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Text;

namespace RefresherSample
{
    public class Refresher : ICacheLoader
    {
        private string _connectionString;
        private static ICache _cache;

        public void Init(IDictionary<string, string> parameters, string cacheName)
        {

            if (parameters != null && parameters.Count > 0)
                _connectionString = parameters.ContainsKey("ConnectionString")
                    ? parameters["ConnectionString"] as string : string.Empty;
            if (string.IsNullOrEmpty(cacheName))
                throw new ArgumentException("Cache name cannot be null or empty");

            _cache = CacheManager.GetCache(cacheName);
        }

        public object LoadDatasetOnStartup(string dataset)
        {
            IList<object> loadDatasetAtStartup;

            if (string.IsNullOrEmpty(dataset))
                throw new InvalidOperationException("Invalid dataset.");

            switch (dataset.ToLower())
            {
                case "products":
                    loadDatasetAtStartup = FetchProductsFromDataSouce();
                    break;
                case "Orders":
                    loadDatasetAtStartup = FetchOrdersFromDataSouce();
                    break;
                default:
                    throw new InvalidOperationException("Invalid Dataset.");
            }

            string[] keys = GetKeys(loadDatasetAtStartup);
            IDictionary<string, CacheItem> cacheData = GetCacheItemDictionary(keys, loadDatasetAtStartup);
            _cache.InsertBulk(cacheData);

            object userContext = DateTime.Now;
            return userContext;
        }

        public object RefreshDataset(string dataSet, object userContext)
        {
            if (string.IsNullOrEmpty(dataSet))
                throw new InvalidOperationException("Invalid dataset.");

            DateTime? lastRefreshTime;

            switch (dataSet.ToLower())
            {
                case "products":
                    lastRefreshTime = userContext as DateTime?;
                    IList<Product> productsNeedToRefresh = FetchUpdatedProducts(lastRefreshTime) as IList<Product>;
                    foreach(var product in productsNeedToRefresh)
                    {
                        string key = $"ProductID:{product.Id}";
                        CacheItem cacheItem = new CacheItem(product);
                        _cache.Insert(key, cacheItem);
                    }
                    break;
                case "Orders":
                    lastRefreshTime = userContext as DateTime?;
                    IList<Order> OrdersNeedToRefresh = FetchUpdatedOrders(lastRefreshTime) as IList<Order>;
                    foreach (var order in OrdersNeedToRefresh)
                    {
                        string key = $"OrderID:{order.OrderID}";
                        CacheItem cacheItem = new CacheItem(order);
                        _cache.Insert(key, cacheItem);
                    }
                    
                    break;
                default:
                    throw new InvalidOperationException("Invalid Dataset.");
            }

            userContext = DateTime.Now;
            return userContext;
        }

        public IDictionary<string, RefreshPreference> GetDatasetsToRefresh(IDictionary<string, object> userContexts)
        {
            IDictionary<string, RefreshPreference> DatasetsNeedToRefresh = new Dictionary<string, RefreshPreference>();
            DateTime? lastRefreshTime;
            bool datasetHasUpdated;

            foreach (var dataSet in userContexts.Keys)
            {
                switch (dataSet.ToLower())
                {
                    case "products":
                        lastRefreshTime = userContexts[dataSet] as DateTime?;
                        datasetHasUpdated = HasProductDatasetUpdated(dataSet, lastRefreshTime);
                        if (datasetHasUpdated)
                        {
                            DatasetsNeedToRefresh.Add(dataSet, RefreshPreference.RefreshNow);
                        }
                        break;
                    case "Orders":
                        lastRefreshTime = userContexts[dataSet] as DateTime?;
                        datasetHasUpdated = HasOrdersDatasetUpdated(dataSet, lastRefreshTime);
                        if (datasetHasUpdated)
                        {
                            DatasetsNeedToRefresh.Add(dataSet, RefreshPreference.RefreshOnNextTimeOfDay);
                        }
                        break;
                    default:
                        throw new InvalidOperationException("Invalid Dataset.");
                }               
            }

            return DatasetsNeedToRefresh;
        }

        public void Dispose()
        {
            // clean your unmanaged resources
        }

        public string[] GetKeys(IList<object> objects)
        {
            string[] keys = new string[objects.Count];
            for (int i = 0; i < keys.Length; i++)
            {
                keys[i] = objects[i].GetType() == typeof(Product) ? $"ProductId:{(objects[i] as Product).Id}" : $"OrderID:{(objects[i] as Order).OrderID}";
            }

            return keys;
        }

        public static IDictionary<string, CacheItem> GetCacheItemDictionary(string[] keys, IList<Object> value)
        {
            IDictionary<string, CacheItem> items = new Dictionary<string, CacheItem>();
            CacheItem cacheItem = null;

            for (int i = 0; i < value.Count - 1; i++)
            {
                cacheItem = new CacheItem(value[i]);
                items.Add(keys[i], cacheItem);
            }

            return items;
        }

        private IList<object> FetchProductsFromDataSouce()
        {
            string Query = "select * from Products where UnitsInStock > 0";
            return ExecuteQuery(Query, "products");

        }

        private IList<object> FetchOrdersFromDataSouce()
        {
            string Query = "select * from dbo.Orders";
            return ExecuteQuery(Query, "Orders");

        }

        private bool HasProductDatasetUpdated(string dataSet, object dateTime)
        {
            bool result = false;
            string query = $"select count(*) from Products where CreationTime > '{dateTime as DateTime?}' and LastModify > '{dateTime as DateTime?}'";
            result = ExecuteAggregateQuery(query) > 0;
            return result;
        }

        private bool HasOrdersDatasetUpdated(string dataSet, object dateTime)
        {
            bool result = false;
            string query = $"select count(*) from Orders where OrderDate > '{dateTime as DateTime?}' and RequiredDate > '{dateTime as DateTime?}'";
            result = ExecuteAggregateQuery(query) > 0;
            return result;
        }
       
        private IList<object> FetchUpdatedProducts(object dateTime)
        {
            string Query = $"select * from Products where CreationTime > '{dateTime as DateTime?}' and LastModify > '{dateTime as DateTime?}'";
            return ExecuteQuery(Query, "products");

        }

        private IList<object> FetchUpdatedOrders(object dateTime)
        {           
            string Query = $"select * from Products where CreationTime > '{dateTime as DateTime?}' and LastModify > '{dateTime as DateTime?}'";
            return ExecuteQuery(Query, "products");

        }     
        
        private int ExecuteAggregateQuery(string Query)
        {
            int result = 0;
            using (SqlConnection myConnection = new SqlConnection(_connectionString))
            {
                SqlCommand oCmd = new SqlCommand(Query, myConnection);

                myConnection.Open();
                using (var reader = oCmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        result = Convert.ToInt32(reader[0]);
                    }                   
                }
                myConnection.Close();
            }
            return result;
        }
        
        private IList<object> ExecuteQuery(string Query, string dataSet)
        {
            IList<object> Data;
            
            using (SqlConnection myConnection = new SqlConnection(_connectionString))
            {                
                SqlCommand oCmd = new SqlCommand(Query, myConnection);

                myConnection.Open();

                using (var reader = oCmd.ExecuteReader())
                {
                    Data = GetData(reader, dataSet);
                }
                    
                myConnection.Close();
            }            
            return Data ;
        }

        private IList<object>  GetData(SqlDataReader sqlDataReader, string dataSet)
        {
            IList<object> dataList = new List<object>();
            while (sqlDataReader.Read())
            {
                if (string.Compare(dataSet, "Orders", true) == 0)
                {
                    Order orders = new Order()
                    {
                        OrderID = Convert.ToInt32(sqlDataReader["OrderID"]),
                        ShipName = sqlDataReader["ShipName"].ToString(),
                        ShipAddress = sqlDataReader["ShipAddress"].ToString(),
                        ShipCity = sqlDataReader["ShipCity"].ToString()
                    };
                    dataList.Add(orders);

                }
                if (string.Compare(dataSet, "Products", true) == 0)
                {
                    Product product = new Product()
                    {
                        Name = sqlDataReader["ProductName"].ToString(),
                        Id = Convert.ToInt32(sqlDataReader["ProductID"]),
                        UnitsAvailable = Convert.ToInt32(sqlDataReader["UnitsInStock"]),
                        UnitPrice = Convert.ToInt32(sqlDataReader["UnitPrice"])
                    };
                    dataList.Add(product);
                    
                }
            }
            return dataList;
        }
        
    }
}
